Blish HUD/GameServices/Input/KeyboardManager.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using Blish_HUD.Input.WinApi;
+using Microsoft.Xna.Framework.Input;
+
+namespace Blish_HUD.Input {
+    public class KeyboardManager : InputManager {
+
+        #region Event Handling
+
+        /// <summary>
+        /// Occurs when a key is pressed down.
+        /// </summary>
+        public event EventHandler<KeyboardEventArgs> KeyPressed;
+
+        /// <summary>
+        /// Occurs when a key is released.
+        /// </summary>
+        public event EventHandler<KeyboardEventArgs> KeyReleased;
+
+        /// <summary>
+        /// Occurs when the state of any key changes.
+        /// </summary>
+        public event EventHandler<KeyboardEventArgs> KeyStateChanged;
+
+        private void OnKeyStateChanged(KeyboardEventArgs e) {
+            if (e.EventType == KeyboardEventType.KeyDown) {
+                this.KeyPressed?.Invoke(this, e);
+            } else {
+                this.KeyReleased?.Invoke(this, e);
+            }
+
+            this.KeyStateChanged?.Invoke(this, e);
+        }
+
+        #endregion
+
+        // TODO: Block using a stack of contexts which can independently process incoming keyboard events
+        private bool _hookGeneralBlock;
+
+        /// <summary>
+        /// The current state of the keyboard.
+        /// </summary>
+        public KeyboardState State { get; private set; }
+
+        /// <summary>
+        /// A <see cref="ModifierKeys"/> flag indicating the acting
+        /// modifier keys (Ctrl, Alt, or Shift) being pressed.
+        /// </summary>
+        public ModifierKeys ActiveModifiers { get; private set; }
+
+        private readonly ConcurrentQueue<KeyboardEventArgs> _inputBuffer;
+
+        private readonly List<Keys> _keysDown;
+
+        /// <summary>
+        /// A list of keys currently being pressed down.
+        /// </summary>
+        public IReadOnlyList<Keys> KeysDown => _keysDown.AsReadOnly();
+
+        internal KeyboardManager() : base(HookType.WH_KEYBOARD_LL) {
+            _keysDown    = new List<Keys>();
+            _inputBuffer = new ConcurrentQueue<KeyboardEventArgs>();
+        }
+
+        internal override void Update() {
+            while (_inputBuffer.TryDequeue(out var keyboardEvent)) {
+                _keysDown.Remove(keyboardEvent.Key);
+
+                if (keyboardEvent.EventType == KeyboardEventType.KeyDown) {
+                    _keysDown.Add(keyboardEvent.Key);
+                }
```

Comment:
the preview here is not showing everything. you remove something and then add it again shortly after

Reply:
[Link to Discord discussion](https://discordapp.com/channels/531175899588984842/536970543736291346/653335593635282945)

Summary: my current implementation is a bit hard to follow.  Benchmarks show there is also a negligible impact on performance (0.572μs processing 242 simulated key state changes) between this and the alternative.  In the interesting of making things easier to maintain, I'll be swapping this over to use a full if/else instead of relying on the conditional inside of the lists `Remove` function.